network.h

```diff
@@ -246,219 +245,250 @@ class network
             mutate_activation(*this, mut_rate_act, rng);
             mutate_weights(*this, mut_rate_weight, mut_step, rng);
             mut_dupl_node(*this, mut_rate_dup, rng);
             mut_del(*this, mut_rate_dup, rng);
-          }
+        }
 
-          else if constexpr(M == mutation_type::NRaddition)
-          {
+        else if constexpr(M == mutation_type::NRaddition)
+        {
             mutate_biases(*this, mut_rate_weight, mut_step, rng);
             mutate_activation(*this, mut_rate_act, rng);
             mutate_weights(*this, mut_rate_weight, mut_step, rng);
             mut_add_node(*this, mut_rate_dup, rng);
             mut_del(*this, mut_rate_dup, rng);
-          }
+        }
 
-        };
+    };
 
     inline std::vector<node>::iterator get_empty_node_in_layer(size_t l)
     {
-     std::vector<node> &layer = get_net_weights()[l];
-     return std::find_if(layer.begin(), layer.end(), node_is_inactive);
+        std::vector<node> &layer = get_net_weights()[l];
+        return std::find_if(layer.begin(), layer.end(), node_is_inactive);
     }
 
 
-    inline void duplicate_node(const node &to_duplicate, size_t layer, size_t index_to_duplicate,
-                                    const std::vector<node>::iterator &empty_node_iterator)
+    inline void duplicate_node(const node &node_to_duplicate,
+                               size_t layer,
+                               size_t index_to_duplicate,
+                               const std::vector<node>::iterator &empty_node_iterator)
     {
-      if (layer == m_current_arc.size() - 1) throw "help :(";
-        if(m_current_arc.at(layer+1) >= m_max_arc.at(layer+1))
+        if (layer == (m_current_arc.size() - 1))
+        {
+            throw std::runtime_error{"help :(, trying to duplicate one node in output"};
+        }
+
+        //This is bad, very confusing
+        //Check if layer has free nodes, if not just return
+        //Should be already checked in mutate_duplication
+        if(m_current_arc.at(layer + 1) == m_max_arc.at(layer + 1))
+        {
             return;
+        }
 
         size_t index = empty_node_iterator - get_net_weights().at(layer).begin() ;
-        m_network_weights.at(layer).at(index) = to_duplicate;
+        m_network_weights.at(layer).at(index) = node_to_duplicate;
 
+        //This also should trigger if first check fails
         for(auto &node : m_network_weights.at(layer+1)){
             weight weight_to_duplicate = node.get_vec_weights().at(index_to_duplicate);
             node.change_nth_weight(weight_to_duplicate, index);
         }
         ++m_current_arc.at(layer+1);
+
+        if(m_current_arc.at(layer + 1) > m_max_arc.at(layer + 1))
+        {
+            std::runtime_error{"One node too much has been added"};
+        }
     }
 
-    inline void add_node(size_t layer, const std::vector<node>::iterator &empty_node_iterator, std::mt19937_64 &rng)
+    inline void add_node(size_t layer,
```

Comment:
This function does not seem to give problems, but should definitely be rewritten by wrapping some of its content into smaller function that should be tested themselves. Not now ofc, but in the future...